src/thor/expansion_action.cc

```diff
@@ -0,0 +1,183 @@
+#include "thor/worker.h"
+
+#include "baldr/json.h"
+#include "baldr/rapidjson_utils.h"
+#include "midgard/constants.h"
+#include "midgard/logging.h"
+#include "midgard/polyline2.h"
+#include "midgard/util.h"
+
+using namespace rapidjson;
+using namespace valhalla::midgard;
+
+namespace {
+// these need to represent array types
+enum PropType {
+  kEdgeId,
+  kStatus,
+  kDist,
+  kTime,
+  kCost,
+  // insert new keys before this
+  LAST
+};
+
+static std::unordered_map<PropType, const char*> kPropPaths =
+    {{PropType::kEdgeId, "/features/0/properties/edge_ids"},
+     {PropType::kStatus, "/features/0/properties/statuses"},
+     {PropType::kDist, "/features/0/properties/distances"},
+     {PropType::kTime, "/features/0/properties/durations"},
+     {PropType::kCost, "/features/0/properties/costs"}};
+} // namespace
+
+namespace valhalla {
+namespace thor {
+
+std::string thor_worker_t::expansion(Api& request) {
+  // time this whole method and save that statistic
+  measure_scope_time(request, "thor_worker_t::expansion");
+
+  // get the request params
+  auto options = request.options();
+  auto exp_action = options.expansion_action();
+  bool skip_opps = options.skip_opposites();
+  std::unordered_set<baldr::GraphId> opp_edges;
+
+  // default generalization to ~ zoom level 15
+  float gen_factor = options.has_generalize() ? options.generalize() : 21.4f;
```

Comment:
you'd need to visualize this in qgis or other desktop software, at least my browser hopelessly crashes even for the tiniest isochrone in geojson.io :sweat_smile: which reminds me that we're always adding 10 mins/kms to each dijkstra computation, so there's actually no way to get less than 10 mins/kms expansion (fine for me, but good to keep in mind)

A few stats/pics for different epsilon from Vienna 1 km (aka 11 km) isodistance:
- 10 m epsilon: 112184 total **vertices**, 3.1 MB response
- 21.4 m epsilon: 114660 total **vertices**, 3.1 MB response

the 10 m default has relatively very few more vertices, no difference in response size and simply looks better (the yellow circled road is fine IMO). the pic (red: 20 m, green: 10 m):
![image](https://user-images.githubusercontent.com/25637358/125826200-75db47d6-5462-46eb-907a-3bb9bd0fc104.png)